File: android/app/src/main/java/com/nexus/satellite/WolManager.java
package com.nexus.satellite;

import android.content.Context;
import android.net.wifi.WifiManager;
import android.util.Log;

import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;

public class WolManager {
    private static final String TAG = "NexusWol";

    public static boolean sendWakeOnLan(Context context, String macStr) {
        if (macStr == null) return false;
        try {
            String cleanMac = macStr.replace(":", "").replace("-", "").trim();
            if (cleanMac.length() != 12) {
                Log.e(TAG, "Invalid MAC address: " + macStr);
                return false;
            }

            byte[] macBytes = new byte[6];
            for (int i = 0; i < 6; i++) {
                macBytes[i] = (byte) Integer.parseInt(cleanMac.substring(i * 2, i * 2 + 2), 16);
            }

            byte[] magicPacket = new byte[102];
            for (int i = 0; i < 6; i++) {
                magicPacket[i] = (byte) 0xFF;
            }
            for (int i = 0; i < 16; i++) {
                System.arraycopy(macBytes, 0, magicPacket, 6 + i * 6, 6);
            }

            WifiManager wifiManager = (WifiManager) context.getApplicationContext().getSystemService(Context.WIFI_SERVICE);
            WifiManager.MulticastLock multicastLock = null;
            if (wifiManager != null) {
                multicastLock = wifiManager.createMulticastLock("NexusWolMulticastLock");
                multicastLock.acquire();
            }

            DatagramSocket socket = new DatagramSocket();
            socket.setBroadcast(true);

            // 1. General Broadcast 255.255.255.255 (Ports 9 & 7)
            InetAddress genBroadcast = InetAddress.getByName("255.255.255.255");
            socket.send(new DatagramPacket(magicPacket, magicPacket.length, genBroadcast, 9));
            socket.send(new DatagramPacket(magicPacket, magicPacket.length, genBroadcast, 7));

            // 2. Subnet Broadcast 192.168.100.255 (Ports 9 & 7)
            try {
                InetAddress subnetBroadcast = InetAddress.getByName("192.168.100.255");
                socket.send(new DatagramPacket(magicPacket, magicPacket.length, subnetBroadcast, 9));
                socket.send(new DatagramPacket(magicPacket, magicPacket.length, subnetBroadcast, 7));
            } catch (Exception ignored) {}

            socket.close();

            if (multicastLock != null && multicastLock.isHeld()) {
                multicastLock.release();
            }

            Log.i(TAG, "WOL magic packet broadcasted for " + macStr);
            return true;
        } catch (Exception e) {
            Log.e(TAG, "Error sending WOL: " + e.getMessage(), e);
            return false;
        }
    }
}

File: android/app/src/main/java/com/nexus/satellite/WolManager.kt
package com.nexus.satellite

import android.content.Context
import android.net.wifi.WifiManager
import android.util.Log
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress

object WolManager {
    private const val TAG = "NexusWol"

    fun sendWakeOnLan(context: Context, macStr: String): Boolean {
        return try {
            val cleanMac = macStr.replace(":", "").replace("-", "").trim()
            if (cleanMac.length != 12) {
                Log.e(TAG, "Invalid MAC address: $macStr")
                return false
            }

            val macBytes = ByteArray(6)
            for (i in 0 until 6) {
                macBytes[i] = cleanMac.substring(i * 2, i * 2 + 2).toInt(16).toByte()
            }

            val magicPacket = ByteArray(102)
            for (i in 0 until 6) {
                magicPacket[i] = 0xFF.toByte()
            }
            for (i in 0 until 16) {
                System.arraycopy(macBytes, 0, magicPacket, 6 + i * 6, 6)
            }

            // Acquire WiFi Multicast lock if on Android
            val wifiManager = context.applicationContext.getSystemService(Context.WIFI_SERVICE) as? WifiManager
            val multicastLock = wifiManager?.createMulticastLock("NexusWolMulticastLock")
            multicastLock?.acquire()

            val broadcastAddr = InetAddress.getByName("255.255.255.255")
            val socket = DatagramSocket()
            socket.broadcast = true

            // Send to port 9
            val packet9 = DatagramPacket(magicPacket, magicPacket.size, broadcastAddr, 9)
            socket.send(packet9)

            // Send to port 7 (fallback)
            val packet7 = DatagramPacket(magicPacket, magicPacket.size, broadcastAddr, 7)
            socket.send(packet7)

            socket.close()
            multicastLock?.release()

            Log.i(TAG, "WOL magic packet broadcasted for $macStr")
            true
        } catch (e: Exception) {
            Log.e(TAG, "Error sending WOL: ${e.message}", e)
            false
        }
    }
}
